### PR TITLE
[EncoderDecoder] Add encoder-decoder for roberta/ vanilla longformer

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR adds Roberta to the Encoder Decoder framework. Thus, it automatically makes it possible to use both `Roberta2Roberta` models and `Longformer2Roberta` model: 

```python
from transformers import EncoderDecoderModel
model = EncoderDecoderModel.from_pretrained("roberta-base", "roberta-base")
input_ids = torch.tensor([10 * [0]])
model(input_ids=input_ids, decoder_input_ids=input_ids)
```

and 

```python
from transformers import EncoderDecoderModel
model = EncoderDecoderModel.from_pretrained("allenai/longformer-base-4096", "roberta-base")
input_ids = torch.tensor([10 * [0]])
model(input_ids=input_ids, decoder_input_ids=input_ids)
```

Also pinging @ibeltagy and @patil-suraj

### CODE REVIEW

File: src/transformers/modeling_bert.py [hunk 1]
```diff
@@ -956,7 +956,7 @@ def __init__(self, config):
         super().__init__(config)
 
         if not config.is_decoder:
-            logger.info("If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`")
+            logger.warning("If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`")
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:959]: @LysandreJik - I changed it to a warning becasue it's actually wrong IMO to use `BertLMHeadModel` with bi-directional self-attention

File: tests/test_modeling_bert.py [hunk 2]
```diff
@@ -152,7 +152,7 @@ def prepare_config_and_inputs_for_decoder(self):
             encoder_attention_mask,
         )
 
-    def create_and_check_bert_model(
+    def create_and_check_model(
```
Comment (patrickvonplaten) [inline tests/test_modeling_bert.py:155]: shorten name -> no need to add `_bert` here because all tests are only relevant for BERT

File: src/transformers/modeling_encoder_decoder.py [hunk 3]
```diff
@@ -191,11 +191,9 @@ def forward(
         input_ids=None,
         inputs_embeds=None,
         attention_mask=None,
-        head_mask=None,
```
Comment (patrickvonplaten) [inline src/transformers/modeling_encoder_decoder.py:194]: The values should only optionally be put in via `kwargs` since not every model has a `head_mask` argument (Longformer does not).

File: src/transformers/modeling_roberta.py [hunk 4]
```diff
@@ -139,6 +145,14 @@ def create_position_ids_from_inputs_embeds(self, inputs_embeds):
             Optionally, instead of passing :obj:`input_ids` you can choose to directly pass an embedded representation.
             This is useful if you want more control over how to convert `input_ids` indices into associated vectors
             than the model's internal embedding lookup matrix.
+        encoder_hidden_states  (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_length, hidden_size)`, `optional`, defaults to :obj:`None`):
```
Comment (sgugger) [inline src/transformers/modeling_roberta.py:148]: Are those arguments used in any model beside the causal LM one? If not, maybe they should only be exposed/documented there (with `labels`).

File: src/transformers/modeling_roberta.py [hunk 5]
```diff
@@ -175,6 +189,116 @@ def set_input_embeddings(self, value):
...
+        position_ids=None,
+        head_mask=None,
+        inputs_embeds=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        labels=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+        **kwargs
```
Comment (sgugger) [inline src/transformers/modeling_roberta.py:229]: Are you using those kwargs? If so change the docstrings since there is no legacy arguments here.

File: tests/test_modeling_bert.py (hunk 2 above)
Comment (sgugger) [reply tests/test_modeling_bert.py:155]: Completely agree with this renaming.

**Review by sgugger — APPROVED** [review]
Great to see the encoder/decoder framework expanded! Thanks for all the work!

**Review by sshleifer — APPROVED** [review]
LGTM. Have people had good ROUGE with the compose two pretrained glue models and finetune for summarization approach?

File: src/transformers/modeling_roberta.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_roberta.py]: Removed them from the corresponding BERT model as well.

File: src/transformers/modeling_roberta.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_roberta.py]: changed it in BERT as well.

**Comment by patrickvonplaten** [comment]
> LGTM. Have people had good ROUGE with the compose two pretrained glue models and finetune for summarization approach?

Hmm, I think it's very new so not sure if many people have tried out the framework yet. @patil-suraj - do you know if people work a lot with EncoderDecoder by chance?

**Comment by patil-suraj** [comment]
> do you know if people work a lot with EncoderDecoder by chance?

Seems like it, seen quite a few issues and questions (on forum as well) regarding EncoderDecoder, but no one has reported any good results yet

**Comment by ibeltagy** [comment]
Looks great. Thanks, @patrickvonplaten. 

> LGTM. Have people had good ROUGE with the compose two pretrained glue models and finetune for summarization approach?
@sshleifer, was thinking about the same thing. My guess is that numbers won't be great because cross-attention is randomly initialized?

File: src/transformers/modeling_bert.py (hunk 1 above)
Comment (LysandreJik) [reply src/transformers/modeling_bert.py:961]: You're right, it's better.

File: src/transformers/modeling_roberta.py [hunk 6]
```diff
@@ -175,6 +181,123 @@ def set_input_embeddings(self, value):
         self.embeddings.word_embeddings = value
 
 
+@add_start_docstrings(
+    """RoBERTa Model with a `language modeling` head on top for CLM fine-tuning. """, ROBERTA_START_DOCSTRING
+)
+class RobertaForCausalLM(BertPreTrainedModel):
```
Comment (LysandreJik) [inline src/transformers/modeling_roberta.py:187]: Wouldn't it be more coherent to have it as `RobertaLMHeadModel`?

Comment (LysandreJik) [reply src/transformers/modeling_roberta.py:187]: (But I do prefer `RobertaForCausalLM`)

File: tests/test_modeling_bert.py (hunk 2 above)
Comment (LysandreJik) [reply tests/test_modeling_bert.py:155]: Nice!

File: src/transformers/modeling_bert.py [hunk 7]
```diff
@@ -976,22 +978,27 @@ def forward(
         position_ids=None,
         head_mask=None,
         inputs_embeds=None,
-        labels=None,
         encoder_hidden_states=None,
         encoder_attention_mask=None,
+        labels=None,
```
Comment (LysandreJik) [inline src/transformers/modeling_bert.py:983]: good catch

File: src/transformers/modeling_roberta.py [hunk 8]
```diff
@@ -175,6 +181,123 @@ def set_input_embeddings(self, value):
...
+            is used in the cross-attention if the model is configured as a decoder.
+            Mask values selected in ``[0, 1]``:
+            ``1`` for tokens that are NOT MASKED, ``0`` for MASKED tokens.
+        labels (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`, defaults to :obj:`None`):
+            Labels for computing the left-to-right language modeling loss (next word prediction).
+            Indices should be in ``[-100, 0, ..., config.vocab_size]`` (see ``input_ids`` docstring)
+            Tokens with indices set to ``-100`` are ignored (masked), the loss is only computed for the tokens with labels
+            in ``[0, ..., config.vocab_size]``
+        kwargs (:obj:`Dict[str, any]`, optional, defaults to `{}`):
+            Used to hide legacy arguments that have been deprecated.
```
Comment (LysandreJik) [inline src/transformers/modeling_roberta.py:237]: No need for this I think

**Review by LysandreJik — APPROVED** [review]
Great, LGTM!

File: src/transformers/modeling_roberta.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/modeling_roberta.py:187]: It's the same terminology for BERT.

Comment (patrickvonplaten) [reply src/transformers/modeling_roberta.py:187]: Following internal discussion will leave the name as it is more precise and `BertLMHeadModel` should change in the future.

File: src/transformers/modeling_roberta.py [hunk 9]
```diff
@@ -175,6 +181,123 @@ def set_input_embeddings(self, value):
...
+            Mask to avoid performing attention on the padding token indices of the encoder input. This mask
+            is used in the cross-attention if the model is configured as a decoder.
+            Mask values selected in ``[0, 1]``:
+            ``1`` for tokens that are NOT MASKED, ``0`` for MASKED tokens.
+        labels (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`, defaults to :obj:`None`):
+            Labels for computing the left-to-right language modeling loss (next word prediction).
+            Indices should be in ``[-100, 0, ..., config.vocab_size]`` (see ``input_ids`` docstring)
+            Tokens with indices set to ``-100`` are ignored (masked), the loss is only computed for the tokens with labels
+            in ``[0, ..., config.vocab_size]``
+        kwargs (:obj:`Dict[str, any]`, optional, defaults to `{}`):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_roberta.py:236]:
```suggestion
```

File: src/transformers/modeling_roberta.py (hunk 8 above)
Comment (patrickvonplaten) [inline src/transformers/modeling_roberta.py:237]:
```suggestion
```

**Comment by patrickvonplaten** [comment]
> Looks great. Thanks, @patrickvonplaten.
> 
> > LGTM. Have people had good ROUGE with the compose two pretrained glue models and finetune for summarization approach?
> 
> @sshleifer, was thinking about the same thing. My guess is that numbers won't be great because cross-attention is randomly initialized?

Btw, this paper does some great analysis on reusing checkpoints for Seq2Seq training: https://arxiv.org/pdf/1907.12461.pdf
